naming layers

benanne opened the pull request:
See #115 for discussion.

f0k commented on lasagne/layers/base.py:
```diff
@@ -23,8 +23,10 @@ class Layer(object):
-        self.input_layer = input_layer
+    def __init__(self, input_layer, name=None):
+        if input_layer is not None:
+            self.input_layer = input_layer
```
`self.input_layer` should always exist as a property, shouldn't it? Also changing this will create a merge conflict in #52, so I'd appreciate leaving this line unchanged :)

benanne replied:
No, currently the presence of `self.input_layer` or `self.input_layers` is used by `lasagne.layers.get_all_layers` to traverse the graph: https://github.com/benanne/Lasagne/blob/master/lasagne/layers/helper.py#L57
So `InputLayer` is not supposed to have it.

Of course we could make it so this property exists for `InputLayer`, but is set to `None`. Then we just have to update `get_all_layers` to check for `None`. Either approach is fine for me.

f0k replied:
Ah, I see. It's because `InputLayer` didn't have to call the super constructor before. But then we'll have the same problem with `MultipleInputsLayer`, which shouldn't have a `self.input_layer` member either. Maybe `InputLayer` and `MultipleInputsLayer` shouldn't call the super constructor, but take care of `name` by themselves? Or we need a common base class for `Layer`, `MultipleInputsLayer` and `InputLayer` that takes care of this?

The current solution is a bit dangerous because it means whenever I pass `None` as the first argument of a `DropoutLayer`, for example, it will end up without an `input_layer` property; instead this should be caught and lead to an exception.

benanne replied:
Duplicating the logic for `name` in `InputLayer` kind of smells imo, which is why I modified it so it would call the constructor now. But clearly that isn't an ideal situation either.

Either way catching `input_layer is None` and raising an exception is not something we can do in the `Layer` base class, because the `InputLayer` either doesn't have this property, or it should be `None` (depending on which solution we go for).

But we can't really go around and put that check in every `Layer` subclass either. An intermediate subclass for layers with an input_layer is a possibility, but that complicates things quite a bit and I'm not sure if we want that.

f0k commented on lasagne/layers/base.py:
```diff
@@ -189,28 +190,33 @@ def create_param(param, shape):
-            return theano.shared(param)
+        if name is not None:
+            if self.name is not None:
+                name = "%s.%s" % (self.name, name)
```
If this isn't a static method any more, I need to change back some of the tests in #52. This will remind me of it.

f0k commented on lasagne/layers/base.py:
```diff
@@ -189,28 +190,33 @@ def create_param(param, shape):
-            # initialized correctly yet.
+            # initialized correctly yet. Note that we cannot assign
+            # a name here.
             return param
```
While you're at it, you might want to add a check for `param.ndim == len(shape)`. While we cannot check the shape, we can at least check the dimensionality.

benanne replied:
Good call.

f0k commented:
Looks good to me except for one minor comment (the first one) and one possible addition (the last one)!

benanne commented:
I went ahead and modified the code so that all layer instances now have either `input_layer` or `input_layers`. For `InputLayer`, `input_layer` is set to `None`. `lasagne.layers.get_all_layers` has been updated to check for this case.

f0k commented:
OK, reading the code again, I'd suggest to not have `InputLayer` call the super constructor and take care of the name itself, similar to how `MultipleInputsLayer` does it. I'd also suggest the InputLayer to name the input variable `name.input_var` if it was not provided by the user. A common base class for the three classes (`Layer`, `MultipleInputsLayer` and `InputLayer`, as per [my earlier comment](https://github.com/benanne/Lasagne/pull/117#discussion_r23697732)) would be cleaner, but this can still be refactored if we find we change the constructor more often. After this change, `InputLayer` wouldn't have an `input_layer` member any more, similar to how a `MultipleInputsLayer` doesn't have an `input_layer` member. The tree traversal code could still check for `input_layer = None` if that's what we want for #52.

benanne commented:
I got rid of the last commit and modified `InputLayer` to handle the name attribute itself. I still think it's a bit smelly but all things considered it's probably the most pragmatic solution. And it's only one line of code anyway. If the code for handling this attribute gets any more complicated we will have to refactor this though.

I also changed the input_var name to include the layer name if it exists.

f0k commented:
Nice! +1 for merging.

benanne commented:
I'll merge this tomorrow if nobody else has comments.

dimatura commented:
Looks good as far as I can tell, this will be useful :)
